add function "open_t"

Thibaut-Gudin opened the pull request.

jrochel commented on cordova_in_app_browser.mli:
```diff
@@ -204,3 +203,39 @@ val open_ :
+
+(*[@@@js.implem*)
+
+open Js_of_ocaml.Dom_html
```
Au lieu d'un `open` ...

jrochel commented on cordova_in_app_browser.mli:
```diff
@@ -204,3 +203,39 @@ val open_ :
+val addEventListener :
+  t                                -> (*Where to add the listener*)
+  string        -> (*The event type*)
+  (event -> unit)                  -> (*The call_back*)
```
... essaye `Js_of_ocaml.Dom_html.event` ici.

Thibaut-Gudin replied:
J'avais déjà essayé, le problème restait le même à la seule différence que le message indiquait que "Js_of_ocaml.Dom_html.event_of_js" n'existe pas.
<<
File "cordova_in_app_browser.ml", line 163, characters 47-79:
163 |                                               (Js_of_ocaml.Dom_html.event_of_js
                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound value Js_of_ocaml.Dom_html.event_of_js
make: *** [Makefile:17: build] Error 2
>>

jrochel commented on cordova_in_app_browser.mli:
```diff
@@ -204,3 +203,39 @@ val open_ :
+val addEventListener :
+  t                                -> (*Where to add the listener*)
+  string        -> (*The event type*)
+  (event -> unit)                  -> (*The call_back*)
```
@dannywillems Hello Danny !
Would you know whether/how one can use a type like `Js_of_ocaml.Dom_html.event` in gen_js_api ?

dannywillems replied:
I don't know if I understand correctly the question, but if the function is expecting an event, simply using `Js_of_ocaml.Dom_html.event` should be enough shouldn't?

jrochel replied:
Unfortunately things are not that simple. Simply using `Js_of_ocaml.Dom_html.event` will result in `gen_js_api` generating `Js_of_ocaml.Dom_html.event_of_js`.

So it seems that one would need to supply a transformation function. I've tried all kinds of stuff already, but to no avail yet.

dannywillems replied:
Using Ojs.t?

dannywillems replied:
Haven't tried, ugly, and not sure it is the best idea:
```
type event = Ojs.t

[@@@js.stop]

val of_dom_html : Js_of_ocaml.Dom_html.event -> event

val to_dom_html : event -> Js_of_ocaml.Dom_html.event

[@@@js.start]

[@@@js.implem
let of_dom_html x = Obj.magic x

let to_dom_html x = Obj.magic x]

val addEventListener : t -> string -> (event -> unit) -> unit [@@js.call]
```

dannywillems commented:
First, thanks for your contribution! The OCaml ecosystem changed quite a lot since the beginning of this project (dune, ocamlformat, merlin and OPAM are more mature, etc). I'd like to update to the new ecosystem gradually, and add CI (GitHub actions).
I started a branch with the new tools: https://github.com/dannywillems/ocaml-cordova-plugin-inappbrowser/pull/3. I haven't got time to check if it is fully retro compatible (building the example app for instance). Would there be one of you keen to try it out pls?

jrochel commented:
Will do!

jrochel commented:
@Thibaut-Gudin Will do!

dannywillems commented:
> @Thibaut-Gudin Will do!

Thanks! Very helpful!

Thibaut-Gudin commented:
I tried to import "#3" with the BeSport app and it seems that their is no problem with the compilation.

I'm not entirely sure if I made it correctly so I will precise what I done in order to testing with the #3 in local:
   First I made a local clone of this branch (https://github.com/dannywillems/ocaml-cordova-plugin-inappbrowser/tree/refacto).
   Then I moved in the local clone and I enter those command lines :
        -> opam install ocamlformat.0.15.0 merlin
	-> opam install -y .
	-> opam pin add cordova-plugin-inappbrowser https://github.com/dannywillems/ocaml-cordova-plugin-inappbrowser.git
	-> opam reinstall -y cordova-plugin-inappbrowser

Then I tried to compile everything and the compilation seems to succeed.

jrochel commented:
> -> opam pin add cordova-plugin-inappbrowser [dannywillems/ocaml-cordova-plugin-inappbrowser.git](https://github.com/dannywillems/ocaml-cordova-plugin-inappbrowser.git)

This would make more sense:
```
opam pin add cordova-plugin-inappbrowser dannywillems/ocaml-cordova-plugin-inappbrowser.git#refacto
```

jrochel commented:
Also: `opam install -ydannywillems/ocaml-cordova-plugin-inappbrowser.git .` implies `opam pin .` so you first installed the version you cloned and then dannywillems/ocaml-cordova-plugin-inappbrowser.git which should be identical to the version you cloned.

Thibaut-Gudin commented:
> opam pin add cordova-plugin-inappbrowser dannywillems/ocaml-cordova-plugin-inappbrowser.git#refacto

I tried this command line but the cmd return an error:

<<
[NOTE] Package cordova-plugin-inappbrowser is currently pinned to git+https://github.com/dannywillems/ocaml-cordova-plugin-inappbrowser.git (version 1.0).
[ERROR] Could not synchronize /home/thibaut/.opam/4.11.0/.opam-switch/sources/cordova-plugin-inappbrowser from
        "git+file:///home/thibaut/forks_Js_Unsafe/pull_Danny_Williams/ocaml-cordova-plugin-inappbrowser/dannywillems/ocaml-cordova-plugin-inappbrowser.git#refacto":
        "/usr/bin/git fetch -q" exited with code 128
[ERROR] Error getting source from git+file:///home/thibaut/forks_Js_Unsafe/pull_Danny_Williams/ocaml-cordova-plugin-inappbrowser/dannywillems/ocaml-cordova-plugin-inappbrowser.git#refacto:
          - git+file:///home/thibaut/forks_Js_Unsafe/pull_Danny_Williams/ocaml-cordova-plugin-inappbrowser/dannywillems/ocaml-cordova-plugin-inappbrowser.git#refacto
>>

jrochel commented:
```
opam pin add cordova-plugin-inappbrowser git@github.com:dannywillems/ocaml-cordova-plugin-inappbrowser.git#refacto
```

Thibaut-Gudin commented:
I already test all the function I added on practice, they all seems to works, I tried to add commentary to explain what I added, however I haven't add anything in the "README".

Thibaut-Gudin commented:
Sorry, I realise  their maybe something that I had to add later in the library.

dannywillems commented:
Is it ready to be reviewed?

jrochel commented on README.md:
````diff
@@ -56,3 +56,33 @@ Cordova_in_app_browser.ai_clear_cache true] in
 i#open_ "https://ocaml.org" (Cordova_in_app_browser.target_self) opt
 ```
+
+The function "open_t" is hte same as "open" but it return a "typed"
````
the

Thibaut-Gudin replied:
See this pull-request for the correction and if others corrections are needed for the README:
https://github.com/besport/ocaml-cordova-plugin-inappbrowser/pull/3

jrochel replied:
(@dannywillems No, it's not yet ready to be reviewed yet.)

jrochel commented:
@dannywillems Alright, I'd say this PR is ready for review.

jrochel commented:
Actually no, the commits should be squashed first. (@Thibaut-Gudin)

Thibaut-Gudin commented:
> Actually no, the commits should be squashed first. (@Thibaut-Gudin)

See https://github.com/besport/ocaml-cordova-plugin-inappbrowser/pull/4
I don't know if it will do the trick...
